### PR TITLE
github action: add top images v6 format conversion and check

### PR DESCRIPTION
**Pull request by yawqi**
Add top images v6 format conversion and check in the nightly
top images conversion test.

Signed-off-by: Qi Wang <wangqi@linux.alibaba.com>

### CODE REVIEW

**Comment by yawqi** [comment]
https://github.com/yawqi/image-service/actions/runs/2492069958
https://github.com/yawqi/image-service/actions/runs/2491791134
The CI and nightly conversion have passed in my repo.

**Review by jiangliu — APPROVED** [review]

File: .github/workflows/convert.yml [hunk 1]
```diff
@@ -30,13 +31,35 @@ jobs:
...
+            echo "converting $I:latest to $I:nydus-nightly-v6"
             sudo DOCKER_CONFIG=$HOME/.docker nydusify convert \
                  --source $I:latest \
-                 --target ${{ env.REGISTRY }}/${{ env.ORGANIZATION }}/$I:nydus-latest \
-                 --build-cache ${{ env.REGISTRY }}/${{ env.ORGANIZATION }}/nydus-build-cache:$I
+                 --target ${{ env.REGISTRY }}/${{ env.ORGANIZATION }}/$I:nydus-nightly-v6 \
+                 --build-cache ${{ env.REGISTRY }}/${{ env.ORGANIZATION }}/nydus-build-cache:$I-v6 \
+                 --fs-version 6
             sudo rm -rf ./tmp
+            nydusify check \
```
Comment (imeoer) [inline .github/workflows/convert.yml:61]: We'd better add `--source $SOURCE_IMAGE_REF` option to check that the source image is consistent with the target image.

Comment (yawqi) [reply .github/workflows/convert.yml:61]: Done.

Comment (yawqi) [reply .github/workflows/convert.yml:61]: When I add the `--source` arg, the convert in my repo won't pass.
https://github.com/yawqi/image-service/runs/6875187357?check_suite_focus=true
Haven't found out why.

Comment (yawqi) [reply .github/workflows/convert.yml]: Turns out when specifying the `--source`, the nydusify will use docker to pull the source image, therefore we need to add sudo here.

**Comment by yawqi** [comment]
It seems that when specifying `--source` the nydusify will crash due to the fs-version and `cache/direct` mode setting, still needs some more refinement...

**Comment by imeoer** [comment]
We need to ensure that the test is as complete as possible, so let's resolve all problems before merging it.

**Comment by liubogithub** [comment]
@imeoer With `--source` being added, is there any other concern about this pr?

**Comment by yawqi** [comment]
> @imeoer With `--source` being added, is there any other concern about this pr?

Actually, for current release, we can't add the --source, related to PR#494.

**Comment by jiangliu** [comment]
#493 has been merged

**Comment by imeoer** [comment]
`nydusify check` do validation by the following rules:

- if `--target` be specified, check the nydus image's manifest format;
- if `--source` be specified, pull the source image and compare the consistency of fs metadata between the ociv1 image and the nydus image;
- if `--backend-type registry/oss --backend-config ...` be specified, check fs data between the ociv1 image and the nydus image;

So we'd better add `--source` and `--backend-type registry` options to validate fs consistency for rafs v5/v6.

**Comment by yawqi** [comment]
I will test it in my own repo first because the current release doesn't support specifying `--source` for RAFS v6.

**Comment by liubogithub** [comment]
Now that current release does not support `--source`,  can we move on and merge it? @imeoer

**Comment by imeoer** [comment]
> Now that current release does not support `--source`, can we move on and merge it? @imeoer

This test only ensures that the nydus image manifest is valid (by `--target`) and the bootstrap (v5/v6) format is fine (by fsck), but it doesn't guarantee that the filesystem consistency of source (ociv1) and target (nydus) images are as expected.

The option `--source` has been supported by `nydusify check`, I think this test will cover many real images, we'd better ensure we have resolved all possible problems.

But indeed we can merge this PR first, at least to make sure the nydus format is fine, agree we can move on.

File: .github/workflows/convert.yml [hunk 2]
```diff
@@ -24,19 +25,42 @@ jobs:
...
+      - name: Convert RAFS v5 images
+        run: |
+          for I in $(cat ${{ env.IMAGE_LIST_PATH }}); do
+            echo "converting $I:latest to $I:nydus-nightly-v5"
+            sudo DOCKER_CONFIG=$HOME/.docker nydusify convert \
+                 --source $I:latest \
+                 --target ${{ env.REGISTRY }}/${{ env.ORGANIZATION }}/$I:nydus-nightly-v5 \
+                 --build-cache ${{ env.REGISTRY }}/${{ env.ORGANIZATION }}/nydus-build-cache:$I-v5 \
+                 --fs-version 5
+            sudo rm -rf ./tmp
```
Comment (imeoer) [inline .github/workflows/convert.yml:50]: Maybe we can put a `nydusify check --target ...` first.

**Comment by yqleng1987** [comment]
@yawqi , Your pull request has been updated. A new test job has been submitted. Please wait in patience.

**Comment by yqleng1987** [comment]
@yawqi , 您好，您无权进行此项操作。

File: .github/workflows/convert.yml (hunk 2 above)
Comment (yawqi) [reply .github/workflows/convert.yml:50]: Done. Tested in my repo, https://github.com/yawqi/image-service/actions/runs/2519103389

**Comment by jiangliu** [comment]
Ready for merging?

**Comment by yawqi** [comment]
> Ready for merging?

Yes, thanks! ; )

**Pull request merged**